lighthouse-core/gather/computed/first-contentful-paint.js

```diff
@@ -0,0 +1,137 @@
+      );
+    });
+
+    return dependencyGraph.cloneWithRelationships(node => {
+      if (node.endTime > fcp) return false;
+      // Include EvaluateScript tasks for blocking scripts
+      if (node.type === Node.TYPES.CPU) return node.isEvaluateScriptFor(blockingScriptUrls);
+      // Include non-script-initiated network requests with a render-blocking priority
+      return node.hasRenderBlockingPriority() && node.initiatorType !== 'script';
+    });
+  }
+
+  /**
+   * @param {!Node} dependencyGraph
+   * @param {!TraceOfTabArtifact} traceOfTab
+   * @return {!Node}
+   */
+  static getPessimisticGraph(dependencyGraph, traceOfTab) {
+    const fcp = traceOfTab.timestamps.firstContentfulPaint;
+    const blockingScriptUrls = FirstContentfulPaint.getScriptUrls(dependencyGraph, node => {
+      return node.endTime <= fcp && node.hasRenderBlockingPriority();
+    });
+
+    return dependencyGraph.cloneWithRelationships(node => {
+      if (node.endTime > fcp) return false;
+      // Include EvaluateScript tasks for blocking scripts
+      if (node.type === Node.TYPES.CPU) return node.isEvaluateScriptFor(blockingScriptUrls);
+      // Include all network requests that had render-blocking priority (even script-initiated)
+      return node.hasRenderBlockingPriority();
+    });
+  }
+
+  static async computeLantern(data, artifacts) {
+    const {trace, devtoolsLog} = data;
+    const graph = await artifacts.requestPageDependencyGraph({trace, devtoolsLog});
+    const traceOfTab = await artifacts.requestTraceOfTab(trace);
+    const networkAnalysis = await artifacts.requestNetworkAnalysis(devtoolsLog);
+
+    const optimisticGraph = FirstContentfulPaint.getOptimisticGraph(graph, traceOfTab);
+    const pessimisticGraph = FirstContentfulPaint.getPessimisticGraph(graph, traceOfTab);
+
+    const options = {...networkAnalysis, ...data.throttling};
+    const optimisticEstimate = new Simulator(optimisticGraph, options).simulate().timeInMs;
+    const pessimisticEstimate = new Simulator(pessimisticGraph, options).simulate().timeInMs;
+
+    const timing =
+      COEFFICIENTS.intercept +
+      COEFFICIENTS.optimistic * optimisticEstimate +
+      COEFFICIENTS.pessimistic * pessimisticEstimate;
+
+    return {
+      timing,
+      optimisticEstimate,
+      pessimisticEstimate,
+      optimisticGraph,
+      pessimisticGraph,
+    };
+  }
+
+  /**
+   * @param {{trace: Object, devtoolsLog: Object, throttling: Object}} data
+   * @return {Object}
+   */
+  async compute_(data, artifacts) {
+    if (data.throttling.method !== 'lantern') {
+      const traceOfTab = await artifacts.requestTraceOfTab(data.trace);
+      return {
+        timing: traceOfTab.timings.firstContentfulPaint,
+        timestamp: traceOfTab.timestamps.firstContentfulPaint,
+      };
+    }
+
+    return FirstContentfulPaint.computeLantern(data, artifacts);
+  }
+}
+
+/**
+ * @typedef MetricResult
+ * @property {number} timing
+ * @property {number|undefined} timestamp
```

Comment:
i know it's just an example but shouldn't this be `@property {number=} timestamp`

Reply:
nah it should be more like `@property {number} [timestamp]` [docs](http://usejsdoc.org/tags-property.html) but point taken yes will spend more time on the real typedef when trying to land 😛